### ISSUE TITLE
bmv2 backend does not produce correct JSON for "advanced" header stack usage

### ISSUE THREAD

**Issue by antoninbas** [p4lang/p4c#321]
The bmv2 backend produces an invalid JSON for the attached P4_16 program. bmv2 crashes when the produces JSON is loaded.
Recent additions to bmv2 mean that this P4_16 program is now supported. See https://github.com/p4lang/behavioral-model/pull/292/files#diff-418314966c57fea403ba29bf6b0793c7.
Specifically the following expression is not translated properly: `m.offset + h.hs.last.f2`.
The JSON should look like this:
```
{
  "type": "expression",
  "value": {
    "type": "expression",
    "value": {
      "op": "+",
      "left": {
        "type": "field",
        "value": [
          "meta",
          "offset"
        ]
      },
      "right": {
        "type": "expression",
        "value": {
          "op": "access_field",
          "left": {
            "type": "expression",
            "value": {
              "op": "dereference_stack",
              "left": {
                "type": "header_stack",
                "value": "stack"
              },
              "right": {
                "type": "expression",
                "value": {
                  "op": "last_stack_index",
                  "left": null,
                  "right": {
                    "type": "header_stack",
                    "value": "stack"
                  }
                }
              }
            }
          },
          "right": 1
        }
      }
    }
  }
}
```

[stack_complex.p4.txt](https://github.com/p4lang/p4c/files/791933/stack_complex.p4.txt)

**Comment by antoninbas** [comment]
I'm taking a stab at this, I have some time today...

**Comment by mbudiu-vmw** [comment]
Was planning to work on it tomorrow. Today I am at the iovisor summit.

**Comment by antoninbas** [comment]
@mbudiu-vmw I don't mind doing this one if you can take care of #297

**Comment by antoninbas** [comment]
Initially I was planning on fixing this as described above, but it required treating `hs[last].f` differently in parser select expressions and in assignments. I didn't find an easy way in `ExpressionConverter` to detect whether I was in a select expression (couldn't use visitor `Context` information because the visitor is "invoked" on each expression individually and not on the whole IR). So I made a small change to bmv2 to support `stack_field` in a more general way in expressions (not just in parser keys). This can probably be improved when I find a good way of supporting arbitrary expressions in select expressions in bmv2.

**Comment by mbudiu-vmw** [comment]
The change is very simple, I assume most of the complexity is in changes to bmv2.
We could arrange for the compiler to move all complex expressions from select expressions into prior assignments.

**Comment by mbudiu-vmw** [comment]
One solution would be to add a new method to ExpressionConverter to indicate a special context (e.g., inSelectContext(bool)).
But frankly, I like it more if expressions are treated homogeneously everywhere.

**Comment by antoninbas** [comment]
yes I agree with you that expressions should ideally be treated homogeneously, which is why I decided to update bmv2 to support `stack_field` in a more general way

**Comment by antoninbas** [comment]
Fixed by #338
